craft_parts/utils/partition_utils.py

```diff
@@ -118,58 +115,123 @@ def _validate_partition_naming_convention(partitions: Sequence[str]) -> None:
                 ),
             )
 
+        if _is_valid_partition_name(partition):
+            continue
         raise errors.FeatureError(
             message=f"Partition {partition!r} is invalid.",
             details=PARTITION_INVALID_MSG,
         )
 
 
-def _validate_namespace_conflicts(partitions: Sequence[str]) -> None:
-    """Validate conflicts between regular partitions and namespaces.
+def _validate_partitions_conflicts(partitions: Sequence[str]) -> None:
+    """Validate conflicts between partitions.
+
+    Assumes partition names are valid and unique.
+
+    :raises FeatureError: for conflicts
+    """
+    conflictings_partitions = detect_conflicts(partitions)
+
+    if not conflictings_partitions:
+        return
+
+    filtered_conflictings_partitions = unique_conflicts(conflictings_partitions)
+
+    # Raise the full list of collisions
+    lines: list[str] = []
+    for conflictings in filtered_conflictings_partitions.values():
+        conflictings_list = list(conflictings)
+        conflictings_list.sort()
+        lines.append(f"- {str(conflictings_list)[1:-1]}")
+
+    msg = "\n".join(lines)
+
+    raise errors.FeatureError(
+        message=f"Following partitions conflicts:\n{msg}",
+        details="Hyphens and slashes are converted to underscores to associate partitions names with environment variables. 'foo-bar' and 'foo/bar' would result in environment variable FOO_BAR.",
+    )
+
+
+def detect_conflicts(partitions: Sequence[str]) -> dict[str, set[str]]:
+    """Detect and return every conflicts between partitions.
 
-    For example, `foo` conflicts in ['default', 'foo', 'foo/bar'].
-    Assumes partition names are valid.
+    Rules:
+      1: partition name must not conflict with namespace (or sub-namespace) names
+        - `foo` and 'foo/bar' conflicts
+        - `foo/bar` and 'foo/bar/baz' conflicts
+      2: environment variables derived from partition names must not conflicts
+        - `foo/bar-baz` and `foo/bar/baz` conflicts (would be converted to FOO_BAR_BAZ)
+        - `foo/bar/baz-qux` and `foo/bar-baz/qux` conflicts (would be converted to FOO_BAR_BAZ_QUX)
 
-    :raises FeatureError: for namespace conflicts
     """
-    namespaced_partitions: set[str] = set()
-    regular_partitions: set[str] = set()
+    conflictings_partitions: dict[str, set[str]] = defaultdict(set)
+
+    # Rule #1: Detect conflicts between partitions and namespaces partitions
+    for candidate_partition in partitions:
+        for partition in partitions:
```

Comment:
I think `itertool.combinations()` can help here.

Reply:
Yep, thanks! The code is now also checking both "rules" at the same time. 
In some cases translating the partition names may be unnecessary but since we will never deal with many partitions I think that is fine.